OSDP: Add support for Control Panel mode of operation

sidcha opened the pull request:
This is a follow-up patch (to #26965) add support for CP mode and a sample.

CC: @carlescufi @jukkar @pfalcon @galak

sidcha committed:
mgmt/osdp: Refactor PD for upcomming CP changes

Some PD init sequence and other methods are useful in CP mode as well.
This patch refactors those methods to osdp_*(), moves them to common
source files and and exposes them from osdp_common.h.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

sidcha committed:
mgmt/osdp: Add support for CP mode of operation

Following the PD mode implementation (9a91b4ad), this patch adds support
for CP mode of operation in OSDP.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

sidcha committed:
samples: mgmt/osdp: Scope PD sample for upcoming CP sample

Make changes to PD sample so a CP sample can be added in future along
with it.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

sidcha committed:
samples: mgmt/osdp: Add CP Sample

Add a OSDP CP mode sample that sends out a periocic command to a
connected PD. It also demosnstrates key press and card read callback
registration.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

jukkar commented on subsys/mgmt/osdp/src/osdp_cp.c:
```diff
@@ -0,0 +1,888 @@
+#define REPLY_FMT_DATA_LEN             3   /* variable length command */
+#define REPLY_BUSY_DATA_LEN            0
+
+static int osdp_atoi(const char *a, int *i)
```
I suggest that you use `strtol()` instead of this function, as strtol will handle errors just fine.

cbsiddharth replied:
Done.

jukkar commented on subsys/mgmt/osdp/src/osdp_cp.c:
```diff
@@ -0,0 +1,888 @@
+	return 0;
+}
+
+static char *osdp_strsep(char **str, char *sep)
```
Could you separate this function to minimal libc (send a separate PR for that) as it might be useful to other applications too.

cbsiddharth replied:
Raised PR #27618. Will remove the one here once it is merged.

cbsiddharth replied:
#27618 is merged. So osdp_strsep() is replaced by strtok_r().

jukkar commented on subsys/mgmt/osdp/src/osdp_cp.c:
```diff
@@ -0,0 +1,888 @@
+	max_len = sizeof(pd->rx_buf) - pd->rx_buf_len;
+
+	rec_bytes = pd->channel.recv(pd->channel.data, buf, max_len);
+	if (rec_bytes <= 0)	/* No data received */
```
one liners need {}

cbsiddharth replied:
Sorry, force of habit. I'll try to be more careful. Btw, can we have checkpatch warn for this?

jukkar commented on subsys/mgmt/osdp/src/osdp_cp.c:
```diff
@@ -0,0 +1,888 @@
+ *  0: success
+ * -1: error
+ *  1: no data yet
+ *  2: re-issue command
```
Could you use defines for the return values as it is easier for the reader. Also for example errno values could be used for errors.

cbsiddharth replied:
I have fixed this to some extent (where ever there are more than 2 return codes). I think I must do the same to PD mode also. I will add a new commit that will do the same change there.

jukkar reviewed (approved):
LGTM

cbsiddharth commented:
@carlescufi, this is ready for review.

cbsiddharth commented:
Ping?

@carlescufi @pfalcon @galak

sidcha committed:
mgmt/osdp: Refactor struct osdp_cmd members for readability

Some of the names used in `struct osdp_cmd` where directly as in the
specification. Initially it appealed to keep them like that but with
time, a little more consistent naming of members helps if you haven't
read the specification document very recently.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>
